### DISCUSSION TITLE
How do you add multiple animated gifs to a composite animated gif in different positions?

### DISCUSSION THREAD

**Discussion by JohnGoldInc** [SixLabors/ImageSharp#1483] [Q&A] [answered]
I'd like to display a single animated gif composed of individual animated  gifs (all have same number of frames)

This makes a static image composite:
```
        private static Image AddImage( Image original, Image additional, Point location , Size size) {
                var add = additional.Clone(x=>x.Resize(size));
                original.Mutate(x=>x.DrawImage(add, location, 1));
                return original;
        }
```

**Comment by JohnGoldInc** [comment]
original.Frames.AddFrame does not have a position...

**Reply by JimBobSquarePants** [reply]
https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.ImageFrameCollection-1.html#SixLabors_ImageSharp_ImageFrameCollection_1_InsertFrame_System_Int32_SixLabors_ImageSharp_ImageFrame__0__

**Reply by JohnGoldInc** [reply]
> https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.ImageFrameCollection-1.html#SixLabors_ImageSharp_ImageFrameCollection_1_InsertFrame_System_Int32_SixLabors_ImageSharp_ImageFrame__0__

InsertFrame does not have a x,y position

**Reply by JimBobSquarePants** [reply]
That’s because all frames should be the same dimensions.

**Reply by INeedADip** [reply]
We always do our best to create the smallest Animated GIF possible, and have had great success adding frames that are not the same size. We try to only "cover" the previous image with the change using the "Left" and "Top" properties when the DisposalMethod property of the previous frame is left in place "DisposalMethod.None".

We are attempting to move from GraphicsMill to Imagesharp.  Here are the Animated GIF properties we currently use:

Top and Left:
https://www.graphicsmill.com/docs/gm/AllMembers_T_Aurigma_GraphicsMill_Codecs_GifFrameSettings.htm

DisposalMthod:
https://www.graphicsmill.com/docs/gm/T_Aurigma_GraphicsMill_Codecs_DisposalMethod.htm

**Reply by JimBobSquarePants** [reply]
Yep, so we don't currently support frames of different dimensions so your Top, Left should be translated to the Point used when drawing your image on the frame. 

LZW usually compresses very well with areas of transparency since it's just a count of repeated content so there's no real loss in output size. 

We have the [`GifDisposalMethod`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Formats.Gif.GifDisposalMethod.html) available to use. 

If you use [GifColorTableMode.Global](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Formats.Gif.GifMetadata.html#SixLabors_ImageSharp_Formats_Gif_GifMetadata_ColorTableMode) the quantizer will build a palette by sampling each frame to keep it as small as possible.

**Comment by JohnGoldInc** [comment, marked as answer]
I was hoping this would work, but it doesn't make a valid gif:

```
  private static Image AddImage(Image original, Image additional, Point location, Size size, string filetype)
        {
            if (filetype == "gif")
                for (var f = 0; f < frames; f++)
                {
                    var currentFrame = original.Frames.CloneFrame(0);
                    if (original.Frames.Count() > f)
                        currentFrame = original.Frames.CloneFrame(f);

                    var add = additional.Frames.CloneFrame(f).Clone(x => x.Resize(size));

                    currentFrame.Mutate(x => x.DrawImage(add, location, 1));

                    if (original.Frames.Count > f + 1)
                    {
                        original.Frames.RemoveFrame(f);
                        original.Frames.InsertFrame(f, currentFrame.Frames[0]);
                    }
                    else
                    {
                        original.Frames.AddFrame(currentFrame.Frames[0]);
                    }
                }
            else
            {
                var add = additional.Clone(x => x.Resize(size));
                original.Mutate(x => x.DrawImage(add, location, 1));
            }

            return original;
        }
```

**Reply by JimBobSquarePants** [reply]
Please define “a valid gif”. Any file saved using the gif encoder will be valid to the specification. 

I’m assuming you mean that the animation does not behave in a way you expect. You have to define per frame rules via GifFrameMetadata.

**Reply by JohnGoldInc** [reply]
Figured it out, yes you are right 

This fixed me up 

` tree.Metadata.GetGifMetadata().ColorTableMode = GifColorTableMode.Local;`

I found this before I saw your message:

https://github.com/SixLabors/ImageSharp/issues/942

**Reply by JimBobSquarePants** [reply]
Hurrah! Glad you figured it out. 👍
